[Parrycys] iP

Parrycys opened the pull request:
## Doobert
> “Productivity is never an accident. It is always the result of a commitment to excellence, intelligent planning, and focused effort.” – Paul J. Meyer ([source](https://dansilvestre.com/productivity-quotes/#1_%E2%80%9CProductivity_is_never_an_accident_It_is_always_the_result_of_a_commitment_to_excellence_intelligent_planning_and_focused_effort%E2%80%9D_%E2%80%93_Paul_J_Meyer))

Doobert simplifies your life by tracking your tasks, so you don’t have to constantly worry about due dates. It's
- text-based
- easy to learn
- ~~FAST~~ _SUPER FAST_ to use 

All you need to do is,
1. download it from [here](https://github.com/Parrycys/ip/releases/tag/A-Jar).
2. double-click it.
3. add your tasks. 📃 
4. let it manage your tasks for you 😸 

AND what's best ? It's **FREE**

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon) 

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Doobert {
    public static void main(String[] args) {
            new Doobert("./data/doobert.txt").run();
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Parrycys committed:
Renaming of chatbot, Greeting from chatbot and exit message

Parrycys committed:
Improve skeletal version of Doobert so that it echos commands entered
by the user, and exits when the user types the command "bye".

Parrycys committed:
Add the ability to store whatever text entered by the user and display
them back to the user when requested.

Parrycys committed:
Add the ability to mark tasks as done and add the ability to
change the status back to not done.

Add Task class to keep track of tasks status.

Remove Storage class.

Parrycys committed:
Add support for tracking three types of tasks: ToDos, Deadlines and
Events.

Parrycys committed:
Test using the I/O redirection technique

Parrycys committed:
Teach the chatbot to deal with errors such as incorrect
inputs entered by the user

Parrycys committed:
Add support for deleting tasks from list and add more exception
handling

Parrycys committed:
Save the tasks in the hard disk automatically whenever the task list
changes

Parrycys committed:
Merge branch 'branch-Level-7'

Parrycys committed:
Teach the chatbot how to understand dates and times

Parrycys committed:
Merge branch 'branch-Level-8'

Parrycys committed:
Refactor the code to extract out closely related code as classes

Parrycys committed:
Merge branch 'branch-A-MoreOOP'

Parrycys committed:
Organize the classes into doobert package

Parrycys committed:
Merge branch 'branch-A-Packages'

Parrycys committed:
Merge branch 'add-gradle-support'

Parrycys committed:
Add JUnit tests to test the behavior of the code for AddEventCommmand
and AddDeadlineCommand

Parrycys committed:
Merge branch 'branch-A-JUnit'

Parrycys committed:
Add JavaDoc comments to the code and Tweak the code to comply with
a given coding standard

Parrycys committed:
Merge commit '4f5e19dc16dd5c0269037dbe3c428e1cd53a540a'

Parrycys committed:
Give users a way to find a task by searching for a keyword in the
task description

Parrycys committed:
Merge branch 'branch-A-Level-9'

changjy81 commented on src/main/java/doobert/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,49 @@
+ */
+public class AddDeadlineCommand extends Command {
+    private final String description;
+    private final String by;
```
Could this name be changed to a noun?

changjy81 commented on src/main/java/doobert/AddEventCommand.java:
```diff
@@ -0,0 +1,50 @@
+public class AddEventCommand extends Command {
+
+    private String description;
+    private String from;
+    private String to;
```
Likewise, could these be changed to nouns?

changjy81 commented on src/main/java/doobert/DoobertException.java:
```diff
@@ -0,0 +1,79 @@
+        }
+    }
+
+    public static void fileNotFound(File filePath) throws DoobertException {
```
Perhaps this method name could be a verb instead?

changjy81 commented on src/main/java/doobert/Storage.java:
```diff
@@ -0,0 +1,76 @@
+     *
+     * @param taskList The {@code TaskList} contains the tasks to be saved.
+     */
+    public void saveTask(TaskList taskList) {
```
This method overwrites the entire file and saves the entire list; should this be named as a plural?

changjy81 reviewed (commented):
LGTM overall with some minor nitpicks.

Golddirio commented on src/main/java/doobert/DoobertException.java:
```diff
@@ -0,0 +1,79 @@
+        super(message);
+    }
+
+    public static void validateMarkCommand(int index, int taskListSize) throws DoobertException {
+        if (index < 0 || index >= taskListSize) {
+            throw new DoobertException("Invalid task number. No such task exists.");
+        }
+    }
+
+    public static void validateUnmarkCommand(int index, int taskListSize) throws DoobertException {
+        if (index < 0 || index >= taskListSize) {
+            throw new DoobertException("Invalid task number. No such task exists.");
+        }
+    }
```
May I know what the difference between these two methods?

Golddirio commented on src/main/java/doobert/DoobertException.java:
```diff
@@ -0,0 +1,79 @@
+        }
+    }
+
+    public static void validateDeleteCommand(int index, int taskListSize) throws DoobertException {
+        if (index < 0 || index >= taskListSize) {
+            throw new DoobertException("Invalid task number. No such task exists.");
+        }
+    }
```
Maybe this exception can be combined with the previous two commented exceptions?

Parrycys committed:
Add a GUI to the chatbot. Use the JavaFX technology to implement the GUI

Parrycys committed:
Merge branch 'branch-Level-10'

Parrycys committed:
Use assert feature to document important assumptions that should hold
at various points in the code

Assertions were missing for key assumptions, such as ensuring
non-null descriptions in tasks and verifying valid time formats.

Without these assertions, incorrect inputs could silently cause
unexpected behaviour, making debugging more difficult.

Let's add assertions in constructors and methods to check:
* Task descriptions are not empty or null
* Event and deadline times are correctly formatted
* File creation of doobert.txt is valid

Parrycys committed:
Merge pull request #2 from Parrycys/branch-A-Assertions

Add assertions and code quality

Parrycys committed:
Refactor Event, Deadline and TaskList parsing to improve readability

Parsing logic in Event, Deadline and TaskList was deeply nested and
contained repetitive code, making it difficult to maintain and debug.
Duplicate code was found in DoobertException.

Deep nesting and complex expressions were refactored by:
* Extracting parsing logic into helper methods.
* Using `DateTimeFormatter[]` for cleaner iteration.
* Removing redundant exception handling.

This improves readability, making the code easier to extend in the
future

Parrycys committed:
Add view command to display tasks given a date by user

Previously, there was no way for users to view their tasks in a
structured format, making it difficult to plan and organize upcoming
tasks.

To improve usability, Let's
* Accept a date in dd-MM-yyyy or dd/MM/yyyy format
* Display all tasks scheduled for the given date

Parrycys committed:
Merge branch 'branch-B-ViewSchedules'

Parrycys committed:
Remove methods that are not used and cite ChatGPT for code
generated by AI tools

Parrycys committed:
Update README, user image for chatbot and flexible resizing
for application

Parrycys committed:
Update README

Parrycys committed:
Update README

Parrycys committed:
Update README

Parrycys committed:
Update README to include find and unmark command

Parrycys committed:
Update README formatting for unmark and find commands

Parrycys committed:
Update README to include delete command
